diff --git a/lib/voice/VoiceConnection.js b/lib/voice/VoiceConnection.js
index 85e734c170be325e02d1be30be3ecdf5dc18a370..4e953d99aafad77eacb6f71f33caaa67e9ae6af3 100644
--- a/lib/voice/VoiceConnection.js
+++ b/lib/voice/VoiceConnection.js
@@ -546,7 +546,12 @@ class VoiceConnection extends EventEmitter {
             if(this.connecting || this.ready) {
                 let reconnecting = true;
                 if(code === 4006) {
-                    reconnecting = false;
+                    if (this.channelID) {
+                        reconnecting = true;
+                        err = null;
+                    } else {
+                        reconnecting = false;
+                    }
                 } else if(code === 4014) {
                     if(this.channelID) {
                         data.endpoint = null;
diff --git a/lib/voice/streams/OggOpusTransformer.js b/lib/voice/streams/OggOpusTransformer.js
index b8626456544433d75d63cb13387f0d1c8448a833..447549b7b0853391ac1ef580aa9e2a64dab1c7d5 100644
--- a/lib/voice/streams/OggOpusTransformer.js
+++ b/lib/voice/streams/OggOpusTransformer.js
@@ -69,10 +69,11 @@ class OggOpusTransformer extends BaseTransformer {
         }
     }
 
-    _final() {
-        if(!this.#bitstream) {
+    _final(cb) {
+        if (!this.#bitstream === undefined) {
             this.emit("error", new Error("No Opus stream was found"));
         }
+        cb()
     }
 
     _transform(chunk, enc, cb) {
