diff --git a/lib/voice/VoiceConnection.js b/lib/voice/VoiceConnection.js
index 858a4b60a46388f0647a5a70a44c8c96a9c2ab25..dd4a970abd5c60c99d4079f51c9e39635bbb3172 100644
--- a/lib/voice/VoiceConnection.js
+++ b/lib/voice/VoiceConnection.js
@@ -378,7 +378,12 @@ class VoiceConnection extends EventEmitter {
           return;
         }
         if (code === 4006) {
-          reconnecting = false;
+          if (this.channelID) {
+            reconnecting = true;
+            err = null;
+          } else {
+            reconnecting = false;
+          }
         } else if (code === 4014) {
           if (this.channelID) {
             data.endpoint = null;
diff --git a/lib/voice/streams/OggOpusTransformer.js b/lib/voice/streams/OggOpusTransformer.js
index 8c2aeb78a35c8d3cdf79fa6e21b35ff0b9b4bd6d..24c72f353acb45b517863e5f12cab17f8e4a9de3 100644
--- a/lib/voice/streams/OggOpusTransformer.js
+++ b/lib/voice/streams/OggOpusTransformer.js
@@ -74,7 +74,7 @@ class OggOpusTransformer extends BaseTransformer {
   }
 
   _final(cb) {
-    if (!this._bitstream) {
+    if (!this._bitstream === undefined) {
       this.emit("error", new Error("No Opus stream was found"));
     }
     cb();
