diff --git a/lib/voice/VoiceConnection.js b/lib/voice/VoiceConnection.js
index 20d95abbffe106c303aacb522d06d0a23d6d4860..b04a309822eb81eae73313536277a0dca37ff203 100644
--- a/lib/voice/VoiceConnection.js
+++ b/lib/voice/VoiceConnection.js
@@ -603,8 +603,15 @@ class VoiceConnection extends EventEmitter {
             this.emit("warn", `Voice WS close ${code}: ${reason}`);
             if(this.connecting || this.ready) {
                 let reconnecting = true;
-                if(code === 1000 || code === 4006 || code === 4021 || code === 4017) {
+                if(code === 1000 || code === 4021 || code === 4017) {
                     reconnecting = false;
+                } else if(code === 4006) {
+                    if (this.channelID) {
+                        reconnecting = true;
+                        err = null;
+                    } else {
+                        reconnecting = false;
+                    }
                 } else if(code === 4014 || code === 4022) {
                     if(this.channelID) {
                         data.endpoint = null;
diff --git a/lib/voice/streams/OggOpusTransformer.js b/lib/voice/streams/OggOpusTransformer.js
index fc4016a146d5dc606e53ae44c177536ebcfc2405..075959810dcd1703637777a40f9dc733adbe2798 100644
--- a/lib/voice/streams/OggOpusTransformer.js
+++ b/lib/voice/streams/OggOpusTransformer.js
@@ -70,7 +70,7 @@ class OggOpusTransformer extends BaseTransformer {
     }
 
     _final(cb) {
-        if(!this.#bitstream) {
+        if(this.#bitstream === undefined) {
             this.emit("error", new Error("No Opus stream was found"));
         }
         cb();
